File: max/engine/planner.js
"use strict";

/**
 * planner.js — theory-aware candidate reranking and surface realization.
 *
 * Direct port of python/src/harmony/planner.py (class HarmonyPlanner). The
 * learned models still supply the probability distribution; complexity masks
 * and reranks it, and functional voice-leading is a small tie-breaker. Surface
 * realization happens *after* selection so a session commits exactly the token
 * whose proposal was scored.
 *
 * Python's `random.Random` stream cannot be reproduced in JS, so the caller
 * injects an rng: () -> [0,1) (see rng.js). Parity is therefore asserted over
 * masks, function-bonus math, realize() structure, and distributions — never
 * individual random draws (see planner.test.js).
 */

const {
  parseChord,
  chordComplexityTier,
  complexityLevel,
  diatonicQuality,
  harmonicFunction,
  nearestScaleRoot,
} = require("./theory.js");
const { CANON_ROOT } = require("./chord_vocab.js");

// JS `%` yields a negative result for negative operands; Python's does not.
function mod12(x) {
  return ((x % 12) + 12) % 12;
}

// Reproduces theory.Chord.symbol(): `Root:quality[/Bass]` with canonical roots.
function chordSymbol(rootPc, quality, bassPc) {
  let out = CANON_ROOT[mod12(rootPc)] + ":" + quality;
  if (bassPc !== null && bassPc !== undefined) {
    out += "/" + CANON_ROOT[mod12(bassPc)];
  }
  return out;
}

function clamp(value) {
  const n = typeof value === "number" ? value : Number(value);
  if (!Number.isFinite(n)) {
    return 0.0;
  }
  return Math.max(0.0, Math.min(1.0, n));
}

/**
 * createHarmonyPlanner({key, complexity, gravity, rng}) -> planner.
 *
 * rng is an injected () -> [0,1). Returns an object exposing choose, realize,
 * setKey, setComplexity, setGravity, and a `level` getter.
 */
function createHarmonyPlanner(options) {
  const opts = options || {};
  let key = opts.key === undefined ? "C:maj" : opts.key;
  let complexity = clamp(opts.complexity === undefined ? 0.5 : opts.complexity);
  let gravity = clamp(opts.gravity === undefined ? 0.0 : opts.gravity);
  const rng = opts.rng;
  if (typeof rng !== "function") {
    throw new TypeError("createHarmonyPlanner requires an rng function");
  }

  function level() {
    return complexityLevel(complexity);
  }

  function setKey(value) {
    const t = typeof value === "string" ? value.trim() : "";
    key = t || "C:maj";
  }

  function setComplexity(value) {
    complexity = clamp(value);
  }

  function setGravity(value) {
    gravity = clamp(value);
  }

  function functionBonus(source, target) {
    const a = harmonicFunction(source, key);
    const b = harmonicFunction(target, key);
    let bonus = 0.0;
    if (a === "T" && b === "PD") {
      bonus = 0.18;
    } else if (a === "PD" && b === "D") {
      bonus = 0.28;
    } else if (a === "D" && b === "T") {
      bonus = 0.38 + 0.75 * gravity;
    } else if (a === "T" && b === "D") {
      bonus = 0.08;
    }
    if (a === "C" && b === "C") {
      bonus -= 0.12;
    }
    return bonus;
  }

  /**
   * choose(source, choices, modelName) -> [symbol, prob].
   * Sample once from the model distribution after theory reranking.
   */
  function choose(source, choices, modelName) {
    const name = modelName === undefined ? "model" : modelName;

    let valid = [];
    for (const cand of choices) {
      const symbol = cand[0];
      const prob = Math.max(0.0, Number(cand[1]));
      if (parseChord(symbol)) {
        valid.push([symbol, prob]);
      }
    }
    if (valid.length === 0) {
      if (!choices || choices.length === 0) {
        throw new Error(name + " produced no chord candidates");
      }
      valid = choices.map((cand) => [cand[0], Math.max(0.0, Number(cand[1]))]);
    }

    const lvl = level();
    let permitted = valid.filter((item) => chordComplexityTier(item[0], key) <= lvl);
    // Deterministic relaxation: use the least-complex proposals rather than
    // returning silence when no token sits in the requested tier.
    if (permitted.length === 0) {
      let minimum = Infinity;
      for (const item of valid) {
        const t = chordComplexityTier(item[0], key);
        if (t < minimum) {
          minimum = t;
        }
      }
      permitted = valid.filter((item) => chordComplexityTier(item[0], key) === minimum);
    }

    const target = complexity * 4.0;
    const weighted = [];
    for (const item of permitted) {
      const symbol = item[0];
      const probability = item[1];
      const tier = chordComplexityTier(symbol, key);
      // Model likelihood dominates; this bounded factor nudges candidates
      // toward the chosen tier without inventing transitions.
      const proximity = Math.exp(-0.45 * Math.abs(tier - target));
      const functional = Math.exp(functionBonus(source, symbol));
      const weight = Math.max(probability, 1e-12) * proximity * functional;
      weighted.push([symbol, probability, weight]);
    }

    let total = 0.0;
    for (const w of weighted) {
      total += w[2];
    }
    if (total <= 0) {
      // max(permitted, key=prob); Python's max returns the FIRST max on ties.
      let best = permitted[0];
      for (const item of permitted) {
        if (item[1] > best[1]) {
          best = item;
        }
      }
      return [best[0], best[1]];
    }
    const pick = rng() * total;
    let acc = 0.0;
    for (const w of weighted) {
      acc += w[2];
      if (pick <= acc) {
        return [w[0], w[1]];
      }
    }
    const last = weighted[weighted.length - 1];
    return [last[0], last[1]];
  }

  // Python random.choice(seq): seq[int(random() * len(seq))] behaviorally.
  function choice(seq) {
    return seq[Math.floor(rng() * seq.length)];
  }

  /**
   * realize(symbol) -> symbol. Turn the chosen structural token into the
   * audible v4 chord symbol. Reproduces every tier branch exactly.
   */
  function realize(symbol) {
    const chord = parseChord(symbol);
    if (chord === null) {
      return symbol;
    }
    const lvl = level();

    if (lvl <= 1) {
      let root = chord.rootPc;
      let quality = diatonicQuality(root, key, { seventh: lvl === 1 });
      if (quality === null) {
        root = nearestScaleRoot(root, key);
        quality = diatonicQuality(root, key, { seventh: lvl === 1 });
        if (quality === null) {
          quality = lvl ? "maj7" : "maj";
        }
      }
      // Tier 1 exposes inversion as a controlled structural device. It is
      // deterministic for a fixed seed and leaves roughly half root-position.
      let bass = null;
      if (lvl === 1 && rng() < 0.35) {
        const intervals = { maj7: 4, min7: 3, "7": 4, hdim7: 3, dim7: 3 };
        const interval = quality in intervals ? intervals[quality] : 3;
        bass = mod12(root + interval);
      }
      return chordSymbol(root, quality, bass);
    }

    if (lvl === 2) {
      // Compatibility tier: retain the learned token exactly.
      return chordSymbol(chord.rootPc, chord.quality, chord.bassPc);
    }

    let q = chord.quality;
    const fn = harmonicFunction(chord, key);
    if (lvl === 3) {
      if (fn === "D" || q === "7") {
        q = rng() < 0.5 ? "13" : "9";
      } else if (q.startsWith("min")) {
        q = rng() < 0.35 ? "min11" : "min9";
      } else {
        q = fn === "PD" ? "maj7#11" : "maj9";
      }
      return chordSymbol(chord.rootPc, q, chord.bassPc);
    }

    // Tier 4: altered dominants, upper structures, and occasional tritone
    // substitution.
    let root = chord.rootPc;
    if (fn === "D" || q === "7") {
      if (rng() < 0.22) {
        root = mod12(root + 6);
        q = "13#11";
      } else {
        q = choice(["7b9", "7#9", "7#11", "7b13"]);
      }
    } else if (q.startsWith("min")) {
      q = choice(["min11", "min13"]);
    } else {
      q = choice(["maj7#11", "maj13"]);
    }
    return chordSymbol(root, q, chord.bassPc);
  }

  return {
    choose,
    realize,
    setKey,
    setComplexity,
    setGravity,
    get level() {
      return level();
    },
  };
}

module.exports = { createHarmonyPlanner };

File: versions/v4-theory/python/src/harmony/planner.py
"""Theory-aware candidate reranking and surface realization for v4."""

from __future__ import annotations

import math
import random

from ..chord_vocab import CANON_ROOT
from .theory import (
    Chord,
    chord_complexity_tier,
    complexity_level,
    diatonic_quality,
    harmonic_function,
    nearest_scale_root,
    parse_chord,
)

Candidate = tuple[str, float]


class HarmonyPlanner:
    """Choose among model proposals, then render the selected harmony.

    Candidate probability remains the main signal.  Complexity masks and
    reranks the full distribution; functional voice-leading provides a small
    tie-breaker.  Surface realization happens *after* selection so a neural
    session commits exactly the token whose hidden-state proposal was scored.
    """

    def __init__(
        self,
        *,
        key: str = "C:maj",
        complexity: float = 0.5,
        gravity: float = 0.0,
        seed: int | None = None,
    ) -> None:
        self.key = key
        self.complexity = self._clamp(complexity)
        self.gravity = self._clamp(gravity)
        self._rng = random.Random(seed)

    @staticmethod
    def _clamp(value: float) -> float:
        try:
            return max(0.0, min(1.0, float(value)))
        except (TypeError, ValueError):
            return 0.0

    @property
    def level(self) -> int:
        return complexity_level(self.complexity)

    def set_key(self, value: str) -> None:
        self.key = value.strip() or "C:maj"

    def set_complexity(self, value: float) -> None:
        self.complexity = self._clamp(value)

    def set_gravity(self, value: float) -> None:
        self.gravity = self._clamp(value)

    def _function_bonus(self, source: str, target: str) -> float:
        a = harmonic_function(source, self.key)
        b = harmonic_function(target, self.key)
        bonus = {
            ("T", "PD"): 0.18,
            ("PD", "D"): 0.28,
            ("D", "T"): 0.38 + 0.75 * self.gravity,
            ("T", "D"): 0.08,
        }.get((a, b), 0.0)
        if a == b == "C":
            bonus -= 0.12
        return bonus

    def choose(self, source: str, choices: list[Candidate], model_name: str = "model") -> Candidate:
        """Sample once from the model distribution after theory reranking."""
        valid = [(symbol, max(0.0, float(prob))) for symbol, prob in choices if parse_chord(symbol)]
        if not valid:
            if not choices:
                raise ValueError(f"{model_name} produced no chord candidates")
            valid = [(symbol, max(0.0, float(prob))) for symbol, prob in choices]

        level = self.level
        permitted = [item for item in valid if chord_complexity_tier(item[0], self.key) <= level]
        # Deterministic relaxation: use the least-complex proposals rather than
        # returning silence when a model has no token in the requested tier.
        if not permitted:
            minimum = min(chord_complexity_tier(symbol, self.key) for symbol, _ in valid)
            permitted = [item for item in valid if chord_complexity_tier(item[0], self.key) == minimum]

        target = self.complexity * 4.0
        weighted: list[tuple[str, float, float]] = []
        for symbol, probability in permitted:
            tier = chord_complexity_tier(symbol, self.key)
            # Model likelihood dominates; this bounded factor nudges candidates
            # toward the chosen tier without inventing transitions.
            proximity = math.exp(-0.45 * abs(tier - target))
            functional = math.exp(self._function_bonus(source, symbol))
            weight = max(probability, 1e-12) * proximity * functional
            weighted.append((symbol, probability, weight))

        total = sum(weight for _, _, weight in weighted)
        if total <= 0:
            return max(permitted, key=lambda item: item[1])
        pick = self._rng.random() * total
        acc = 0.0
        for symbol, probability, weight in weighted:
            acc += weight
            if pick <= acc:
                return symbol, probability
        symbol, probability, _ = weighted[-1]
        return symbol, probability

    def realize(self, symbol: str) -> str:
        """Turn the chosen structural token into the audible v4 chord symbol."""
        chord = parse_chord(symbol)
        if chord is None:
            return symbol
        level = self.level

        if level <= 1:
            root = chord.root_pc
            quality = diatonic_quality(root, self.key, seventh=level == 1)
            if quality is None:
                root = nearest_scale_root(root, self.key)
                quality = diatonic_quality(root, self.key, seventh=level == 1) or ("maj7" if level else "maj")
            # Tier 1 exposes inversion as a controlled structural device.  It
            # is deterministic for a fixed seed and leaves roughly half root-position.
            bass = None
            if level == 1 and self._rng.random() < 0.35:
                intervals = {"maj7": 4, "min7": 3, "7": 4, "hdim7": 3, "dim7": 3}
                bass = (root + intervals.get(quality, 3)) % 12
            return Chord(root, quality, bass).symbol()

        if level == 2:
            # This is the compatibility tier: retain the learned token exactly.
            return chord.symbol()

        q = chord.quality
        fn = harmonic_function(chord, self.key)
        if level == 3:
            if fn == "D" or q == "7":
                q = "13" if self._rng.random() < 0.5 else "9"
            elif q.startswith("min"):
                q = "min11" if self._rng.random() < 0.35 else "min9"
            else:
                q = "maj7#11" if fn == "PD" else "maj9"
            return Chord(chord.root_pc, q, chord.bass_pc).symbol()

        # Tier 4: altered dominants, upper structures, and occasional tritone
        # substitution.  The selected model token is still what neural state
        # commits; this is an explicitly audible realization layer.
        root = chord.root_pc
        if fn == "D" or q == "7":
            if self._rng.random() < 0.22:
                root = (root + 6) % 12
                q = "13#11"
            else:
                q = self._rng.choice(("7b9", "7#9", "7#11", "7b13"))
        elif q.startswith("min"):
            q = self._rng.choice(("min11", "min13"))
        else:
            q = self._rng.choice(("maj7#11", "maj13"))
        return Chord(root, q, chord.bass_pc).symbol()
